R/expect_identical_linter.R

```diff
@@ -0,0 +1,80 @@
+#'   equivalent* instead of fully identical (in which case, provide the
+#'   `tolerance=` argument to `expect_equal()` explicitly). This also applies
+#'   when it's inconvenient to check full equality (e.g., names can be ignored,
+#'   in which case `ignore_attr = "names"` should be supplied to
+#'   `expect_equal()` (or, for 2nd edition, `check.attributes = FALSE`).
+#'
+#' @evalRd rd_tags("expect_identical_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+expect_identical_linter <- function() {
+  Linter(function(source_file) {
+    if (length(source_file$parsed_content) == 0L) {
+      return(list())
+    }
+
+    xml <- source_file$xml_parsed_content
+
+    # outline:
+    #   1. conditions for expect_equal()
+    #     - skip when tolerance= is set
+    #     - skip when check.attributes = FALSE [for 2nd edition tests]
+    #     - skip when any ignore_* is set [for 3rd edition tests]
+    #       specifically ignore_{srcref,attr,encoding,function_env,formula_env}
+    #     - skip cases like expect_equal(x, 1.02) or the constant vector version
+    #       where a numeric constant indicates inexact testing is preferable
+    #     - skip calls using dots (`...`); see tests
+    #   2. conditions for expect_true()
+    xpath <- glue::glue("//expr[
+      (
+        SYMBOL_FUNCTION_CALL[text() = 'expect_equal']
+        and not(
+          following-sibling::SYMBOL_SUB[text() = 'tolerance']
+          or following-sibling::SYMBOL_SUB[
+            (
+              text() = 'check.attributes'
+              and following-sibling::expr[1][NUM_CONST[text() = 'FALSE']]
+            )
+            or starts-with(text(), 'ignore_')
+          ]
+          or following-sibling::expr[
+            expr[SYMBOL_FUNCTION_CALL[text() = 'c']]
+            and expr[NUM_CONST[contains(text(), '.')]]
+          ]
+          or following-sibling::expr[NUM_CONST[contains(text(), '.')]]
+          or following-sibling::expr[SYMBOL[text() = '...']]
+        )
+      ) or (
+        SYMBOL_FUNCTION_CALL[text() = 'expect_true']
+        and following-sibling::expr[1][
+          expr[SYMBOL_FUNCTION_CALL[text() = 'identical']]
+        ]
+      )
+    ]")
+
+    bad_expr <- xml2::xml_find_all(xml, xpath)
+    return(lapply(
+      bad_expr,
+      xml_nodes_to_lint,
+      source_file = source_file,
+      message = paste(
+        "expect_identical(x, y) is the default way of comparing two objects in",
+        "testthat unit tests. Only use expect_equal() if testing numeric",
+        "equality (and specifying tolerance= explicitly) or there's a need to",
+        "ignore some attributes, e.g. with ignore_attr = 'names' for the 3rd",
+        "edition or check.attributes = FALSE for the second. For testing",
+        "approximate equality to a whole number, you can also avoid setting",
+        "tolerance= by including an explicit decimal point, e.g.,",
+        "expect_equal(x, 1.0), not expect_equal(x, 1)."
```

Comment:
That's a rather long message compared to other linters.
Maybe shorten to "only use expect_equal() if non-default arguments, e.g. tolerance, are needed"?

Reply:
put some of the details in the man page, should be more digestible now